src/core/proof.ml

```diff
@@ -60,45 +92,125 @@ type proof_state =
 (** Short synonym for qualified use. *)
 type t = proof_state
 
-(** [init builtins name a] returns an initial proof state for a theorem  named
-    [name], which statement is represented by the type [a]. Builtin symbols of
-    [builtins] may be used by tactics, and have been declared. *)
+(** [init name a] returns an initial proof state for a theorem  named
+    [name], which statement is represented by the type [a].
+    The list of goals is not initialized *)
 let init : Pos.strloc -> term -> t = fun name a ->
   let proof_term = fresh_meta ~name:name.elt a 0 in
-  let proof_goals = [Goal.of_meta proof_term] in
-  {proof_name = name; proof_term; proof_goals}
+  {proof_name = name; proof_term; proof_goals = []}
+
+(** [goals_of_meta m] returns a goal associated to the meta m *)
+let goals_of_meta : meta -> Goal.t list = fun m ->
+  let goal_typ = Goal.Typ (Goal.goal_typ_of_meta m) in
+  [goal_typ]
+
+(** [goals_of_typ typ ter] returns a list of goals corresponding to the
+    typability of [typ] by a sort and checking eventually that term
+    [ter] has type [typ] *)
+let goals_of_typ : Pos.popt -> term option -> term option ->
+  Goal.t list * term =
+  fun pos typ ter ->
+  let (typ, sort, to_solve) =
+    match typ, ter with
+    | Some(typ),Some(ter) ->
+      let sort, to_solve2 = Infer.infer [] typ in
+      let to_solve =
+        match sort with
+        | Type | Kind -> Infer.check [] ter typ
+        | _ -> Console.fatal pos "%a has type %a (not a sort)."
+                 pp_term typ pp_term sort
+      in
+      typ, sort, to_solve2 @ to_solve
+    | None,Some(ter) ->
+      let typ, to_solve = Infer.infer [] ter in
+      let sort, to_solve2 =
+        begin
+          match typ with
+          | Kind -> Console.fatal pos "Forbidded definition x := _ -> TYPE"
+          | _ -> let sort, to_solve2 = Infer.infer [] typ in
+            begin
+              match sort with
+              | Type | Kind -> sort,to_solve2
+              | _ -> Console.fatal pos "The term [%a] does not have type \
+                                        [%a]." pp_term ter pp_term typ
```

Comment:
Here, the error message should be that the type of `ter` is not typable by a sort.

Reply:
fixed in 0df87258